File: HardwareTester/services/api_service.py
import requests
from flask import current_app
from requests.adapters import HTTPAdapter
from urllib3.util.retry import Retry
from datetime import datetime
from HardwareTester.utils.custom_logger import CustomLogger
from HardwareTester.extensions import db

# Initialize logger
logger = CustomLogger.get_logger("api_service")

logger.info("Initializing APIService")


class APIService:
    api_state = {
        "initialized": False,
        "running": False,
        "config": {
            "default_machine_name": "Session1",
            "stress_test_mode": False,
            "base_url": None,
            "timeout": 10,
        },
        "blueprints": [],
        "active_sessions": [],
        "endpoints": [],
        "logs": [],
        "errors": [],
        "last_updated": None,
    }

    @staticmethod
    def _get_session_with_retries(retries=3, backoff_factor=0.3, status_forcelist=(500, 502, 504)):
        """
        Configures a session with retry logic for resilient requests.
        """
        session = requests.Session()
        retry = Retry(
            total=retries,
            read=retries,
            connect=retries,
            backoff_factor=backoff_factor,
            status_forcelist=status_forcelist,
        )
        adapter = HTTPAdapter(max_retries=retry)
        session.mount("http://", adapter)
        session.mount("https://", adapter)
        return session

    session = _get_session_with_retries()

    @staticmethod
    def initialize(base_url: str, timeout: int = 10):
        """
        Initialize the APIService with a base URL and timeout.
        """
        if not base_url:
            raise ValueError("Base URL must be provided for initialization.")
        
        APIService.api_state["config"]["base_url"] = base_url
        APIService.api_state["config"]["timeout"] = timeout
        APIService.api_state["initialized"] = True
        APIService.api_state["last_updated"] = datetime.now()
        APIService.api_state["logs"].append(f"APIService initialized at {APIService.api_state['last_updated']}")
        logger.info(f"APIService initialized with base URL: {base_url} and timeout: {timeout}")

    @staticmethod
    def log_error(error_message):
        """
        Log an error in the API state and persist it in logs.
        """
        error_entry = {
            "timestamp": datetime.now(),
            "message": error_message,
        }
        APIService.api_state["errors"].append(error_entry)
        APIService.api_state["last_updated"] = datetime.now()
        logger.error(f"Error logged: {error_message}")

    @staticmethod
    def test_api_connection():
        """
        Test the connection to the API.
        """
        try:
            base_url = APIService.api_state["config"]["base_url"]
            if not base_url:
                raise RuntimeError("APIService is not initialized with a base URL.")

            timeout = APIService.api_state["config"]["timeout"]
            response = APIService.session.get(f"{base_url}/test-connection", timeout=timeout)
            response.raise_for_status()
            logger.info("API connection successful.")
            return {"success": True, "message": "API connection successful"}
        except requests.exceptions.RequestException as e:
            APIService.log_error(f"API connection failed: {e}")
            return {"success": False, "error": f"API connection failed: {e}"}

    @staticmethod
    def fetch_data_from_api(endpoint: str, params: dict = None) -> dict:
        """
        Fetch data from a specified API endpoint.
        """
        try:
            base_url = APIService.api_state["config"]["base_url"]
            if not base_url:
                raise RuntimeError("APIService is not initialized with a base URL.")

            timeout = APIService.api_state["config"]["timeout"]
            response = APIService.session.get(f"{base_url}{endpoint}", params=params, timeout=timeout)
            response.raise_for_status()
            logger.info(f"Data fetched successfully from {endpoint}.")
            return {"success": True, "data": response.json()}
        except requests.exceptions.RequestException as e:
            APIService.log_error(f"Failed to fetch data from {endpoint}: {e}")
            return {"success": False, "error": f"Failed to fetch data from {endpoint}: {e}"}

    @staticmethod
    def push_data_to_api(endpoint: str, payload: dict) -> dict:
        """
        Push data to a specified API endpoint.
        """
        try:
            base_url = APIService.api_state["config"]["base_url"]
            if not base_url:
                raise RuntimeError("APIService is not initialized with a base URL.")

            timeout = APIService.api_state["config"]["timeout"]
            response = APIService.session.post(f"{base_url}{endpoint}", json=payload, timeout=timeout)
            response.raise_for_status()
            logger.info(f"Data pushed successfully to {endpoint}.")
            return {"success": True, "message": "Data successfully pushed"}
        except requests.exceptions.RequestException as e:
            APIService.log_error(f"Failed to push data to {endpoint}: {e}")
            return {"success": False, "error": f"Failed to push data to {endpoint}: {e}"}

    @staticmethod
    def list_available_endpoints() -> dict:
        """
        Fetch the list of available API endpoints.
        """
        try:
            base_url = APIService.api_state["config"]["base_url"]
            if not base_url:
                raise RuntimeError("APIService is not initialized with a base URL.")

            timeout = APIService.api_state["config"]["timeout"]
            response = APIService.session.get(f"{base_url}/endpoints", timeout=timeout)
            response.raise_for_status()
            endpoints = response.json().get("endpoints", [])
            APIService.api_state["endpoints"] = endpoints
            APIService.api_state["last_updated"] = datetime.now()
            logger.info(f"Discovered {len(endpoints)} endpoints.")
            return {"success": True, "endpoints": endpoints}
        except requests.exceptions.RequestException as e:
            APIService.log_error(f"Failed to fetch endpoints: {e}")
            return {"success": False, "error": f"Failed to fetch endpoints: {e}"}

    @staticmethod
    def get_overview() -> dict:
        """
        Fetch a summarized API overview.
        """
        try:
            base_url = APIService.api_state["config"]["base_url"]
            if not base_url:
                raise RuntimeError("APIService is not initialized with a base URL.")

            timeout = APIService.api_state["config"]["timeout"]
            response = APIService.session.get(f"{base_url}/overview", timeout=timeout)
            response.raise_for_status()
            overview_data = response.json()
            logger.info("Successfully fetched API overview.")
            return {"success": True, "overview": overview_data}
        except requests.exceptions.RequestException as e:
            APIService.log_error(f"Failed to fetch API overview: {e}")
            return {"success": False, "error": f"Failed to fetch API overview: {e}"}

File: frontend/src/services/apiService.ts
// src/services/apiService.ts
import axiosInstance from "@/services/axiosInstance";
import { AxiosRequestConfig, AxiosError } from "axios";
import { toast } from "react-toastify";
import { APIState, ErrorEntry, APIResponse } from "@/types/apiTypes";

let apiState: APIState = {
    initialized: false,
    running: false,
    config: {
        default_machine_name: "Session1",
        stress_test_mode: false,
        base_url: null,
        timeout: 10,
    },
    blueprints: [],
    active_sessions: [],
    endpoints: [],
    logs: [],
    errors: [],
    last_updated: null,
};

const logError = (message: string) => {
    const timestamp = new Date().toISOString();
    apiState.errors.push({ timestamp, message });
    apiState.last_updated = timestamp;
    console.error(`[API ERROR] ${message}`);
};

const APIService = {
    /**
     * Initialize the API client.
     */
    initialize(baseUrl: string, timeout: number = 10) {
        if (!baseUrl) throw new Error("Base URL must be provided");
        apiState.config.base_url = baseUrl;
        apiState.config.timeout = timeout;
        apiState.initialized = true;
        apiState.last_updated = new Date().toISOString();
        apiState.logs.push(`Initialized at ${apiState.last_updated}`);
        console.log(`[API INIT] ${baseUrl}`);
    },

    /**
     * Test connection to the backend API.
     */
    async testConnection(): Promise<APIResponse<string>> {
        try {
            await axiosInstance.get("/test-connection");
            return { success: true, data: "API connection successful" };
        } catch (err: any) {
            logError(`API connection failed: ${err.message}`);
            return { success: false, error: `API connection failed: ${err.message}` };
        }
    },

    /**
     * Basic GET with optional query params.
     */
    async fetch<T>(endpoint: string, params?: Record<string, any>): Promise<APIResponse<T>> {
        try {
            const res = await axiosInstance.get<T>(endpoint, { params });
            return { success: true, data: res.data };
        } catch (err: any) {
            logError(`Failed to fetch ${endpoint}: ${err.message}`);
            return { success: false, error: err.message };
        }
    },

    /**
     * Push data to the backend via POST.
     */
    async push(endpoint: string, payload: any): Promise<APIResponse<string>> {
        try {
            await axiosInstance.post(endpoint, payload);
            return { success: true, data: "Data successfully pushed" };
        } catch (err: any) {
            logError(`Failed to push to ${endpoint}: ${err.message}`);
            return { success: false, error: err.message };
        }
    },

    /**
     * List available backend endpoints.
     */
    async listEndpoints(): Promise<APIResponse<string[]>> {
        try {
            const res = await axiosInstance.get("/endpoints");
            const endpoints = res.data?.endpoints || [];
            apiState.endpoints = endpoints;
            apiState.last_updated = new Date().toISOString();
            return { success: true, data: endpoints };
        } catch (err: any) {
            logError(`Failed to list endpoints: ${err.message}`);
            return { success: false, error: err.message };
        }
    },

    /**
     * Get backend summary information.
     */
    async getOverview(): Promise<APIResponse<any>> {
        try {
            const res = await axiosInstance.get("/overview");
            return { success: true, data: res.data };
        } catch (err: any) {
            logError(`Failed to fetch overview: ${err.message}`);
            return { success: false, error: err.message };
        }
    },

    /**
     * Low-level API call wrapper.
     */
    async apiCall<T>(
        url: string,
        method: "GET" | "POST" | "PUT" | "DELETE",
        data: any = null,
        headers: Record<string, string> = {},
        responseType: AxiosRequestConfig["responseType"] = "json",
        timeoutMs?: number
    ): Promise<T> {
        try {
            const config: AxiosRequestConfig = {
                method,
                url,
                data,
                headers,
                responseType,
                timeout: timeoutMs,
            };
            const response = await axiosInstance(config);
            return response.data;
        } catch (error: any) {
            const message = (error as AxiosError)?.message || "API call failed";
            toast.error(`API Error: ${message}`);
            throw error;
        }
    },

    async sleep(ms: number) {
        return new Promise((r) => setTimeout(r, ms));
    },

    /**
     * Compute exponential backoff with full jitter.
     * attempt: 1-based attempt number
     */
    backoffWithJitter(
        baseDelayMs: number,
        attempt: number,
        backoffFactor: number,
        maxDelayMs: number
    ): number {
        const exp = Math.min(maxDelayMs, baseDelayMs * Math.pow(backoffFactor, attempt - 1));
        return Math.floor(Math.random() * exp); // full jitter
    },

    /**
     * Retry-enabled version of apiCall with exponential backoff + Jitter.
     */

    async apiCallWithRetry<T>(
        url: string,
        method: "GET" | "POST" | "PUT" | "DELETE",
        data: any = null,
        headers: Record<string, string> = {},
        retries = 3,
        baseDelayMs = 750,
        backoffFactor = 2,
        maxDelayMs = 10_000,
        baseTimeoutMs?: number,
        maxTimeoutMs = 30_000
    ): Promise<T> {
        const initialTimeoutMs =
            baseTimeoutMs ?? (apiState.config.timeout ? apiState.config.timeout * 1000 : 5_000);

        for (let attempt = 1; attempt <= retries; attempt++) {
            // grow timeout per attempt (capped)
            const attemptTimeoutBase = Math.min(
                maxTimeoutMs,
                initialTimeoutMs * Math.pow(backoffFactor, attempt - 1)
            );
            const attemptTimeoutMs =
                Math.floor(Math.random() * attemptTimeoutBase) || attemptTimeoutBase;

            try {
                // PASS the timeout down to apiCall
                return await APIService.apiCall<T>(
                    url,
                    method,
                    data,
                    headers,
                    "json",
                    attemptTimeoutMs
                );
            } catch (err) {
                if (attempt === retries) throw err;

                // Exponential delay with full jitter before next attempt
                const delay = APIService.backoffWithJitter(baseDelayMs, attempt, backoffFactor, maxDelayMs);
                await APIService.sleep(delay);  // <— FIX: call on APIService
            }
        }
        throw new Error("Unreachable");
    },
    /**
     * Generate the full WebSocket URL based on the configured API base URL.
     * Automatically switches to ws:// or wss:// depending on protocol.
     * @param path - Path to append after /ws/
     */

    getWebSocketURL(path: string): string {
        const baseUrl = apiState.config.base_url || window.location.origin;
        const url = new URL(baseUrl);
        const protocol = url.protocol === "https:" ? "wss:" : "ws:";
        const host = url.host;
        return `${protocol}//${host}/ws/${path}`;
    },
};

export default APIService;
